protocols/bgp/server/bmp_router.go

```diff
@@ -0,0 +1,333 @@
+package server
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	bnet "github.com/bio-routing/bio-rd/net"
+	"github.com/bio-routing/bio-rd/protocols/bgp/packet"
+	bmppkt "github.com/bio-routing/bio-rd/protocols/bmp/packet"
+	"github.com/bio-routing/bio-rd/routingtable"
+	"github.com/bio-routing/bio-rd/routingtable/filter"
+	"github.com/bio-routing/bio-rd/routingtable/locRIB"
+	log "github.com/sirupsen/logrus"
+	"github.com/taktv6/tflow2/convert"
+)
+
+type router struct {
+	address          net.IP
+	port             uint16
+	con              net.Conn
+	reconnectTimeMin int
+	reconnectTimeMax int
+	reconnectTime    int
+	reconnectTimer   *time.Timer
+	rib4             *locRIB.LocRIB
+	rib6             *locRIB.LocRIB
+	neighbors        map[[16]byte]*neighbor
+	neighborsMu      sync.Mutex
+	logger           *log.Logger
+	runMu            sync.Mutex
+	stop             chan struct{}
+}
+
+type neighbor struct {
+	localAS  uint32
+	peerAS   uint32
+	address  [16]byte
+	routerID uint32
+	fsm      *FSM
+	opt      *packet.DecodeOptions
+}
+
+func newRouter(addr net.IP, port uint16, rib4 *locRIB.LocRIB, rib6 *locRIB.LocRIB) *router {
+	return &router{
+		address:          addr,
+		port:             port,
+		reconnectTimeMin: 30,  // Suggested by RFC 7854
+		reconnectTimeMax: 720, // Suggested by RFC 7854
+		reconnectTimer:   time.NewTimer(time.Duration(0)),
+		rib4:             rib4,
+		rib6:             rib6,
+		neighbors:        make(map[[16]byte]*neighbor),
+		logger:           log.New(),
+		stop:             make(chan struct{}),
+	}
+}
+
+func (r *router) serve(con net.Conn) {
+	r.con = con
+	r.runMu.Lock()
+	defer r.con.Close()
+	defer r.runMu.Unlock()
+
+	for {
+		select {
+		case <-r.stop:
+			return
+		default:
+		}
+
+		msg, err := recvBMPMsg(r.con)
+		if err != nil {
+			r.logger.Errorf("Unable to get message: %v", err)
+			return
+		}
+
+		bmpMsg, err := bmppkt.Decode(msg)
+		if err != nil {
+			r.logger.Errorf("Unable to decode BMP message: %v", err)
+			return
```

Comment:
Killing the server because of a stupid neighbor with broken implementation?

Reply:
We're killing the connection to the stupid neighbour and only the stupid neighbour.